### ISSUE TITLE
Software Engineering

### ISSUE THREAD

**Issue by darwinrlo** [darwinrlo/public#3]

**Comment by darwinrlo** [comment]
[Static analysis in GCC 10](https://developers.redhat.com/blog/2020/03/26/static-analysis-in-gcc-10/)

(It'd be pretty awesome if these comments looked up the title of a pasted link and provided the option to replace the link text with it -- just like in Coda.)

**Comment by darwinrlo** [comment]
[Real Not Complex](https://realnotcomplex.com/)

**Comment by darwinrlo** [comment]
[Fabulous adventures in coding: Eric Lippert's blog](https://ericlippert.com/about-eric-lippert/)

> Hello and welcome to my blog! I’m Eric Lippert; I design and implement developer tools at Facebook.

> Before starting at Facebook in February of 2016, I worked on C# static analyzers at Coverity; before that I was a Principal Developer at Microsoft on the C# compiler team and a member of the C# language design team. In my sixteen years at Microsoft I also worked on the design and implementation of VBScript, JScript, Windows Script Host and Visual Studio Tools for Office. I was also once on the ECMA committee which standardizes the JavaScript language.

**Comment by darwinrlo** [comment]
It seems that the proliferation of software engineering interview prep products has increased the depth of the field and consequently raised the bar at top companies like Google and Facebook. This is good for the profession.

**Comment by darwinrlo** [comment]
[The Hardest Program I've Ever Written](http://journal.stuffwithstuff.com/2015/09/08/the-hardest-program-ive-ever-written/)

> I work on the Dart programming language. Part of my job is helping make more Dart code, readable, idiomatic, and consistent, which is why I ended up writing our style guide. That was a good first step, but any style guide written in English is either so brief that it’s ambiguous, or so long that no one reads it.

> Go’s “gofmt” tool showed a better solution: automatically format everything. Code is easier to read and contribute to because it’s already in the style you’re used to. Even if the output of the formatter isn’t great, it ends those interminable soul-crushing arguments on code reviews about formatting.

> Of course, I still have to sell users on running the formatter in the first place. For that, having great output really does matter. Also, I’m pretty picky with the formatting in my own code, and I didn’t want to tell users to use a tool that I didn’t use myself.

> Getting that kind of quality means applying pretty sophisticated formatting rules. That in turn makes performance difficult. I knew balancing quality and speed would be hard, but I didn’t realize just how deep the rabbit hole went.

> I have finally emerged back into the sun, and I’m pleased with what I brought back. I like the output, and the performance is solid. On my laptop, it can blow through over two million lines of code in about 45 seconds, using a single core.

**Comment by darwinrlo** [comment]
[Discover the best YC advice for startup success](https://ycadvice.com/?category=Product)

**Comment by darwinrlo** [comment]
[Ask HN: Computer Science/History Books](https://news.ycombinator.com/item?id=22692281)

Really great recommendations for non-technical books!

**Comment by darwinrlo** [comment]
Possible alternative to algebraic data types.

Define a class `Result<T>`. T is the type of the object that is returned. Methods should return `Result` objects.

Another possibility is for the result of function call to be 

The disadvantage here is that the code deviates from mathematical expressions.

Functions that evaluate mathematical expressions could be a type of function within the language.

Scala is truly the next iteration. Though I haven't looked into Clojure or Kotlin at all.

**Comment by darwinrlo** [comment]
[Effective C: An Introduction to Professional C Programming](https://nostarch.com/Effective_C)

**Comment by darwinrlo** [comment]
[Writing an OS in Rust](https://os.phil-opp.com/)

I like the blogging system that the site owner is using.

Also, this is intriguing:

> The first step in creating our own operating system kernel is to create a Rust executable that does not link the standard library. This makes it possible to run Rust code on the bare metal without an underlying operating system.

**Comment by darwinrlo** [comment]
The mindset you bring to a situation is very important. A mindset that is not a match for the situation will lead to negative emotions that, at best, need to be addressed, adding unnecessary friction. At worst, they are counterproductive and block you from achieving your goals.

**Comment by darwinrlo** [comment]
I can't just read. I need to be seeking to do something. It could be taking notes -- putting things into my own words -- or even just finding things to underline (extracting takeaways).

(This comment made it into the wrong place. I'd like to be able to move this to another issue. When comments are moved between issues, the links to them should be preserved.)

**Comment by darwinrlo** [comment]
[Brilliant: Learn to think](https://brilliant.org/)

**Comment by darwinrlo** [comment]
One function of art is to give expression to feelings. Another function of art is to show what is possible. It can be used to "paint" a vision of reality that would be worthwhile to aspire to.

**Comment by darwinrlo** [comment]
I don't think we need operator syntax (`a op b`) for operations (which are just functions). I think method call syntax is just fine (`a.op(b)`). Though I can see the argument for prefix notation (`(op a b)`), such as used in Lisp, in which case I would personally prefer brackets over parentheses (`[op a b]`). Even better would be just `Op[a, b]`) -- in fact, it'd be pretty sweet to be able to take a list and apply an operator to it.

I get the desire of infix operators for arithmetic expressions, e.g. `1+2`. But I still think prefix notation is better, i.e. `(+)[1, 2]`.

For things that are not operations, such as relationships (which return a boolean value), the infix operator notation makes a lot of sense. Example: `a isSiblingOf b`. For more than two values, prefix notation still makes sense, i.e. `areSiblings(a, b, c)`. Even if it's just two values, `areSiblings(a, b)` would make more sense than `a areSiblings b` -- the pattern here seems to be bidirectional relationships. For relationships (as opposed to operations), I prefer parentheses over brackets. :^)

As much flack as it gets, I do like using postfix and prefix unary operators for incrementing variables, e.g. `++i` and `i++`.

I like camel case for variable names. For functions, I'd actually prefer capitalized names, e.g. `Oldest(darwinLo, johnnieLo)`. Though I like camel case for names of boolean functions, e.g. `areSiblings(a, b)`.

I don't like implicit order of operations. Everything should just be parenthesized.

I'd love to be able to specify a graph declaratively.

```
Node a, b;
Relationship isParentOf, isChildOf;
a isParentOf b implies b isChildOf a;
a isParentOf b;
if b isChildOf a
  Print("Hello!")  // should print!
```

(I like semicolons when you're adding something to the state, such as declaring a new variable or adding a relation.)

Even something like:

```
type Node;
Node a, b;
Relationship isParentOf, isChildOf;
BiRelationship areSiblings;
a isParentOf b iff b isChildOf a;
```

```
a isParentOf b and a isParentOf c implies areSiblings(b, c);
Node a, b, c, d;
a isParentOf b;
a isParentOf c;
a isParentOf d;
areSiblings(b, c, d)  // returns true
```

I do like distinguishing between expressions and statements. Statements should end in a semicolon.

There are two types of if-else structures actually. Sometimes, they're control structures. Sometimes, they're expression evaluators.

x isParentOf b implies b isChildOf a

The following is added.

`a isParentOf b`
`b isChildOf a`

And a relationship can be queried: `a isChildOf b` is an expression that returns `true` or `false`.

The primitive values are very important.

This could be the DSL I need to work on my LeetCode solver.

You can express the recursive substructure (the optimal substructure as the case may be), and the code will generated for you, memoization and all.

```
type Box(width: Int, height: Int, depth: Int);  // A type is an interface.

class Box implements Box {  // A class is an implementation.
}

Relationship <Element> isElementOf <Set>;
e isElementOf l;
a isElement b iff a contains b;  // iff can be used to define synonyms.
```

I don't like getters and setters in Java. Better would be preprocessors and post-processors for properties.

I like the idea of inheritance without subtyping. Maybe have an `inherits` keyword.

Idea: Ranges can also be specified for a given type.

TODO Define a list recursively.

```
<List> :=: <Head> :: <Tail: List>
```

Idea: Keywords are bolded. If it's not bolded, it's not a keyword.

We need to do something about comments. Comments should not be part of the source text. They should be done by the editor. They should collapsible. You should be able to add comments to lines. But you should also be able to annotate specific sub-expressions.

For an editor, something like Workflowy would be great. So you can collapse and expand code. You can add comments. Maybe Lisp is better suited for an editor like Workflowy.

* The language should be tightly integrated with the editor. 
* Everything should be real-time.
* Debugger from the very start.

I like Python's use of whitespace. I don't like Java's braces. It should be indents, not spaces.

Why are we still coding up recursive algorithms by hand?

I do like this construct.

```
x = 3 if y = 5;
```

I think Scala has this.

I do like the use of colons in one-liners, e.g. `if x == 4: x += 1`. And I do like omitting the `return` keyword whether in a one-liner or not, e.g. `if x == 4: x + y`. The value of the last expression of a function body is what is returned.

I wonder if programming languages should support subscripts, superscripts, and so on.

**Comment by darwinrlo** [comment]
I'm really digging the addition of unsigned integers to Java.

**Comment by darwinrlo** [comment]
An expression can be a subtype of `LanguageConstruct`. The language object can be computed at runtime. For example, a function can be computed and returned. Then it can be applied to the expression next to it.

Functions should be "first-class" values. That is, we should be able to pass them to and return them from functions. Objects should also be first-class values.

For some types of problems, functions are the way to go. For others, objects are the way to go.

For some types of problems, expressing something declaratively is the way to go. For others, expressing them procedurally is the way to go.

**Comment by darwinrlo** [comment]
A DSL for problems that can be solved recursively.

- Specify the initial conditions.
- Specify the base case.
- Specify the recursive structure, i.e. how the results from recursive calls are combined.

Excellent test cases
* CtCI 8.11 Coins
* CtCI 8.12 Eight Queens
  * John Whaley used this as a test case for bddbddb.
* CtCI 8.13 Stack of Boxes

The system should automatically infer that an array is sorted.

isSorted

The system can think ahead and see if an array being sorted would help with the algorithm.

**Comment by darwinrlo** [comment]
[Linux Academy](https://linuxacademy.com/)

> Experience interactive courses in AWS, Google Cloud, and Azure and master the tools that shape technology.

**Comment by darwinrlo** [comment]
You can represent summations pretty easily in Scala or Python, e.g. `sum([i+1 for i in range(0, 5)])`.

**Comment by darwinrlo** [comment]
Completing the square

**Comment by darwinrlo** [comment]
[10 Things I Hate About PostgreSQL](https://medium.com/@rbranson/10-things-i-hate-about-postgresql-20dbab8c2791)

**Comment by darwinrlo** [comment]
[The Phoenix Project](https://www.amazon.com/Phoenix-Project-DevOps-Helping-Business/dp/0988262592)

**Comment by darwinrlo** [comment]
* [Security Engineering: A Guide to Building Dependable Distributed Systems](https://www.cl.cam.ac.uk/~rja14/book.html)
* [The Adventures of OS: Making a RISC-V Operating System using Rust](http://osblog.stephenmarz.com/index.html)
* [Crafting Interpreters](http://craftinginterpreters.com/)

**Comment by darwinrlo** [comment]
[Lean Book: The Hitchhiker's Guide to Logical Verification (HN)](https://news.ycombinator.com/item?id=22794533)

**Comment by darwinrlo** [comment]
# Product Management
* [ProductPlan: Everything You Need to Know About Prioritization Frameworks](https://www.productplan.com/product-management-frameworks/)

**Comment by darwinrlo** [comment]
# Software Engineering
* [The Overflow: Best practices for REST API design](https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/)
* [The Overflow: A practical guide to writing technical specs](https://stackoverflow.blog/2020/04/06/a-practical-guide-to-writing-technical-specs/)
* [Fundamentals of Software Architecture: An Engineering Approach](https://www.amazon.com/Fundamentals-Software-Architecture-Engineering-Approach-ebook/dp/B0849MPK73)
* [Software Engineering at Google: Lessons Learned from Programming Over Time](https://www.amazon.com/Software-Engineering-Google-Lessons-Programming-ebook/dp/B0859PF5HB)

**Comment by darwinrlo** [comment]
# Security

* [Nagarro Security: Interactive guide to Buffer Overflow exploitation](https://nagarrosecurity.com/blog/interactive-buffer-overflow-exploitation)

**Comment by darwinrlo** [comment]
# Computer Science
* [Stanford: CS 228](https://cs.stanford.edu/~ermon/cs228/index.html)
* [Stanford: CS 221](https://stanford-cs221.github.io/spring2020/)
* [Stanford: CS 242](https://cs242.stanford.edu/f19/)
* [Stanford: CS 166 - Data Structures](http://web.stanford.edu/class/archive/cs/cs166/cs166.1196/)
* [Keith Schwarz](https://www.keithschwarz.com/)
* [Stanford: CS 103 - Mathematical Foundations of Computing](http://web.stanford.edu/class/cs103//handouts/Mathematical%20Foundations%20of%20Computing.pdf)
* [Stanford: CS 168 - The Modern Algorithmic Toolbox](https://web.stanford.edu/class/cs168/index.html)
* [Stanford: CS 261 - Optimization and Algorithmic Paradigms](http://web.stanford.edu/class/cs261/lectures.html)
* [Stanford: CS 250 - Algebraic Error Correcting Codes](http://web.stanford.edu/~marykw/classes/CS250_W18/index.html#homework)
* [Stanford: CS 265 - Randomized Algorithms and Probabilistic Analysis](http://theory.stanford.edu/~valiant/teaching/CS265/index.html)
* [Stanford 361: Engineering Design Optimization](https://web.stanford.edu/class/aa222/cgi-bin/wordpress/)
* [Stanford 345: Topics in Database Management Systems](https://web.stanford.edu/class/cs345d-01/)

**Comment by darwinrlo** [comment]
# Artificial Intelligence
* [fast.ai -- making neural nets uncool again](https://www.fast.ai/)
* [deeplearning.ai](https://www.deeplearning.ai/)
* [Kaggle](https://www.kaggle.com/)

**Comment by darwinrlo** [comment]
[How Bochs Works Under the Hood](http://bochs.sourceforge.net/How%20the%20Bochs%20works%20under%20the%20hood%202nd%20edition.pdf)

**Comment by darwinrlo** [comment]
[Kaleidoscope: Implementing a Language with LLVM](http://llvm.org/docs/tutorial/index.html)

**Comment by darwinrlo** [comment]
Python is pretty good as pseudocode. I wonder if Julia is even better.
